### PR TITLE
Document most of the remaining parts of the style system, make docs-by-default for the whole style system

### PR DESCRIPTION
**Pull request by emilio**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14819)

### CODE REVIEW

**Comment by emilio** [comment]
BTW, remaining parts are:

 * The `longhands` and `shorthands` modules. My idea for those, given it's mostly parsing code, would be to require just a spec link per property.

 * The parts that are being rewritten (`atomic_refcell` and `media_queries`), because I don't want to waste too much time.

 * The `attr` module.

 * The `gecko` bits. Probably not too hard to do.

**Comment by stshine** [comment]
You are so amazing.

**Comment by Manishearth** [comment]
Reviewed 2 of 2 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8.
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/style/properties/properties.mako.rs, line 486 at r2](https://reviewable.io:443/reviews/servo/servo/14819#-K_Rgh6OsCIkWcIRddyP:-K_Rgh6OsCIkWcIRddyQ:b-6zuhvd) ([raw file](https://github.com/servo/servo/blob/a1a6c27d4ce17cace7fefff5f53771939c2a39c9/components/style/properties/properties.mako.rs#L486)):*
> ```Rust
>     }
> 
>     /// Trie to serialize the given declarations as this shorthand, or return an
> ```

"try"

"return an error if writing to the stream fails. In case the declarations do not map to a shorthand, this will not write anything but will return a success result."

---

*[components/style/properties/properties.mako.rs, line 671 at r2](https://reviewable.io:443/reviews/servo/servo/14819#-K_RhDirzebrJ2Lz4B4-:-K_RhDirzebrJ2Lz4B40:bm0wn7g) ([raw file](https://github.com/servo/servo/blob/a1a6c27d4ce17cace7fefff5f53771939c2a39c9/components/style/properties/properties.mako.rs#L671)):*
> ```Rust
> 
> impl<'a> PropertyDeclarationId<'a> {
>     /// Whether a given declaration id is either the same as `other`,  or a
> ```

double space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14819)*

**Comment by emilio** [comment]
Addressed your nits, went ahead with the Gecko docs too, and fixed the build error that was because I changed the stylist's `viewport_constraints` to return `Option<&ViewportConstraints>` instead of `&Option<ViewportConstraints>`, but I assumed it was only used in style and thus `Ctrl-C`d the build.

I'd appreciate a fast-ish review time, since pretty much any patch that checks in undocumented code in the style system will make me grow this patch.

@bors-servo try

**Comment by emilio** [comment]
cc @heycam @SimonSapin @upsuper and @bzbarsky, which aren't mentioned automatically by bors and haven't been assigned nor reviewed anything.

If you have any objection to this, please speak before landing. Also, review if you feel like it :)

I'm aware the docs for some methods are probably not as extensive as they should, but it's five AM and oh well...

**Comment by Manishearth** [comment]
Reviewed 37 of 37 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/style/parser.rs, line 47 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Tuq39pJvzJ1kvrveM:-K_Tuq39pJvzJ1kvrveN:b-6co8it) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/parser.rs#L47)):*
> ```Rust
> /// The data that the parser needs from outside in order to parse a stylesheet.
> pub struct ParserContext<'a> {
>     /// The `Origin` of the stylesheet.
> ```

Might be better to be explicit here -- "The `Origin` of a stylesheet -- whether it is a user, author, or user-agent sheet"

(`Origin` is easily confused with URL origins)

---

*[components/style/gecko_bindings/sugar/ns_com_ptr.rs, line 5 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Ttk1xFow3PqI4B5sE:-K_Ttk1xFow3PqI4B5sF:b2bpqex) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/gecko_bindings/sugar/ns_com_ptr.rs#L5)):*
> ```Rust
>  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
> 
> //! Litle helpers for `nsCOMPtr`.
> ```

speling

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14819)*

**Comment by Manishearth** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r10, 1 of 1 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14819#-:-K_TyyjzbxnB5dVmTfrE:bnfp4nl)*

**Comment by emilio** [comment]
On Mon, Jan 02, 2017 at 03:25:14AM -0800, Manish Goregaokar wrote:
> *[components/style/parser.rs, line 47 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Tuq39pJvzJ1kvrveM:-K_Tuq39pJvzJ1kvrveN:b-6co8it) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/parser.rs#L47)):*
> > ```Rust
> > /// The data that the parser needs from outside in order to parse a stylesheet.
> > pub struct ParserContext<'a> {
> >     /// The `Origin` of the stylesheet.
> > ```
> 
> Might be better to be explicit here -- "The `Origin` of a stylesheet -- whether it is a user, author, or user-agent sheet"
> 
> (`Origin` is easily confused with URL origins)

Good call, done :)

> *[components/style/gecko_bindings/sugar/ns_com_ptr.rs, line 5 at r9](https://reviewable.io:443/reviews/servo/servo/14819#-K_Ttk1xFow3PqI4B5sE:-K_Ttk1xFow3PqI4B5sF:b2bpqex) ([raw file](https://github.com/servo/servo/blob/caebd5f98501e4563a9998e9ac0a5983e298f4c6/components/style/gecko_bindings/sugar/ns_com_ptr.rs#L5)):*
> > ```Rust
> >  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
> > 
> > //! Litle helpers for `nsCOMPtr`.
> > ```
> 
> speling

Good catch.

**Comment by emilio** [comment]
@bors-servo r=Manishearth

Thanks for the review Manish :)

File: components/style/gecko_bindings/sugar/style_complex_color.rs [hunk 1]
```diff
@@ -17,6 +19,7 @@ impl From<nscolor> for StyleComplexColor {
 }
 
 impl StyleComplexColor {
+    /// Create a `StyleComplexColor` value that represents `currentColor`.
```
Comment (upsuper) [inline components/style/gecko_bindings/sugar/style_complex_color.rs]: [`currentcolor` is speced all-lowercased now](https://drafts.csswg.org/css-color-4/#currentcolor-color).

**Pull request merged**